File: backend-spring-boot/src/test/java/no/vicx/backend/jwt/github/GitHubFromOpaqueProducerTest.java
package no.vicx.backend.jwt.github;

import no.vicx.backend.jwt.github.vm.GitHubUserResponseVm;
import no.vicx.backend.jwt.github.vm.GitHubUserVm;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatus;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.oauth2.server.resource.introspection.BadOpaqueTokenException;
import org.springframework.web.client.HttpClientErrorException;

import java.util.Collections;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class GitHubFromOpaqueProducerTest {

    @Mock
    GitHubUserFetcher userFetcher;

    @InjectMocks
    GitHubFromOpaqueProducer sut;

    @Test
    void createPrincipal_givenValidToken_expectPrincipal() {
        when(userFetcher.fetchUser(anyString())).thenReturn(
                new GitHubUserResponseVm(
                        new GitHubUserVm(
                                "12345",
                                "john-doe",
                                "John Doe",
                                "john.doe@example.com",
                                "https://example.com/avatar.jpg"
                        ),
                        "repo, user",
                        "valid-token"));

        var principal = sut.createPrincipal("valid-token");

        assertNotNull(principal);
        assertEquals("john-doe", principal.getName());
        assertEquals(
                Collections.singletonList("ROLE_GITHUB_USER"),
                principal.getAuthorities().stream().map(GrantedAuthority::getAuthority).toList());
    }

    @Test
    void createPrincipal_givenWebClientResponseException_expectBadOpaqueTokenException() {
        when(userFetcher.fetchUser(anyString())).thenThrow(HttpClientErrorException.create(
                HttpStatus.UNAUTHORIZED,
                HttpStatus.UNAUTHORIZED.getReasonPhrase(),
                null,
                null,
                null));

        var exception = assertThrows(
                BadOpaqueTokenException.class, () -> sut.createPrincipal("valid-token"));

        assertEquals("Invalid or expired GitHub access token", exception.getMessage());
    }

    @Test
    void createPrincipal_givenRuntimeException_expectBadOpaqueTokenException() {
        when(userFetcher.fetchUser(anyString())).thenThrow(new RuntimeException());

        var exception = assertThrows(
                BadOpaqueTokenException.class, () -> sut.createPrincipal("valid-token"));

        assertEquals("Error validating GitHub token", exception.getMessage());
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/jwt/github/GitHubFromOpaqueProducerTest.kt
package no.vicx.backend.jwt.github

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import no.vicx.backend.jwt.github.GitHubTestUtils.githubUserInTest
import no.vicx.backend.jwt.github.vm.GitHubUserResponseVm
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.security.oauth2.server.resource.introspection.BadOpaqueTokenException
import org.springframework.web.client.HttpClientErrorException

class GitHubFromOpaqueProducerTest :
    BehaviorSpec({
        val githubUserFetcher: GitHubUserFetcher = mockk()
        val sut = GitHubFromOpaqueProducer(githubUserFetcher)

        Given("createPrincipal, valid GitHub token response") {
            every { githubUserFetcher.fetchUser(any()) } returns
                GitHubUserResponseVm(
                    user = githubUserInTest,
                    grantedScopes = "repo, user",
                    token = "valid-token",
                )

            When("calling createPrincipal with valid token") {
                val principal = sut.createPrincipal("~valid-token~")

                Then("it should return the principal") {
                    principal.name shouldBe githubUserInTest.login
                }
            }
        }

        Given("createPrincipal, failure scenarios") {
            When("calling createPrincipal with invalid token") {
                every { githubUserFetcher.fetchUser(any()) } throws
                    HttpClientErrorException.create(
                        HttpStatus.UNAUTHORIZED,
                        HttpStatus.UNAUTHORIZED.reasonPhrase,
                        HttpHeaders.EMPTY,
                        "".toByteArray(),
                        null,
                    )

                val thrown =
                    shouldThrow<BadOpaqueTokenException> {
                        sut.createPrincipal("~invalid-token~")
                    }

                Then("exception should contain invalid token message") {
                    thrown.message shouldBe "Invalid or expired GitHub access token"
                }
            }

            When("an unknown exception is thrown when calling createPrincipal") {
                every { githubUserFetcher.fetchUser(any()) } throws RuntimeException("~some-error~")

                val thrown =
                    shouldThrow<BadOpaqueTokenException> {
                        sut.createPrincipal("~valid-token~")
                    }

                Then("exception should contain generic error message") {
                    thrown.message shouldBe "Error validating GitHub token"
                }
            }
        }
    })
